# Ratings Project

Importing the required libraries

In [1]:
import selenium
import pandas as pd
from selenium import webdriver
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
!pip install undetected-chromedriver

  Created wheel for undetected-chromedriver: filename=undetected_chromedriver-3.1.3-py3-none-any.whl size=25769 sha256=d1d390311ee10894e20c8bfaf3e037974b52edc3b34eb6aab5b3a19d18128c32
  Stored in directory: c:\users\sudhanshu kumar\appdata\local\pip\cache\wheels\b3\8d\30\3df6fab41e50613bb19508eaed644023e83183b12ba3977cb4
Successfully built undetected-chromedriver
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.25.11
    Uninstalling urllib3-1.25.11:
      Successfully uninstalled urllib3-1.25.11
  Attempting uninstall: selenium
    Found existing installation: selenium 3.141.0
    Uninstalling selenium-3.141.0:
      Successfully uninstalled selenium-3.141.0


ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

requests 2.24.0 requires urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1, but you'll have urllib3 1.26.8 which is incompatible.


Web Scraping 

In [74]:
import undetected_chromedriver.v2 as uc
driver = uc.Chrome(version_main=95)

In [75]:
url = 'https://www.flipkart.com/boat-airdopes-381-asap-charge-bluetooth-headset/product-reviews/itm350f1ba32ebb0?pid=ACCFMRFXMAGVGAEG&lid=LSTACCFMRFXMAGVGAEGJ0NTWJ&marketplace=FLIPKART&page=1'

In [76]:
driver.get(url)

In [77]:
abc_ = [] #Empty List

In [78]:
for i in range(1,1950):
    try:
        abc = driver.find_elements_by_xpath("//div[@class='_3LWZlK _1BLPMq']")
        for i in abc:
            abc_.append(i.text)
            
    except:abc_.append('-')
        
        
        
btn = driver.find_element_by_xpath("//a[@class='_1LKTO3']")
btn

<selenium.webdriver.remote.webelement.WebElement (session="da1c54b9df6de2d65e90a78b17590ea3", element="829d8da6-0e84-461b-863d-64100f4085dc")>

In [79]:
len(abc_)

19490

In [80]:
xyz_ = [] #Empty List

In [81]:
for i in range(1,1950):
    try:
        xyz = driver.find_elements_by_xpath("//div[@class='t-ZTKy']")
        for i in xyz:
            xyz_.append(i.text)
            
    except:xyz_.append('-')
        
        
        
btn = driver.find_element_by_xpath("//a[@class='_1LKTO3']")
btn

<selenium.webdriver.remote.webelement.WebElement (session="da1c54b9df6de2d65e90a78b17590ea3", element="829d8da6-0e84-461b-863d-64100f4085dc")>

In [82]:
len(xyz_)

19490

In [83]:
import pandas as pd

In [87]:
df = pd.DataFrame()

In [88]:
df["ratings"] = abc_
df["reviews"] = xyz_

In [89]:
df.head(3)

ratings                                            reviews
0       5  Loved the product it's an A grade in every sph...
1       5  The delivery was super fast......\n\nSo writin...
2       5  Bass is excellent 10outof10\nSound is nice 10o...

In [90]:
df["reviews"] = df["reviews"].str.replace(r'^.+@[^\.].*\.[a-z]{2,}$',
                                 'emailaddress')

# Replace URLs with 'webaddress'
df["reviews"] = df["reviews"].str.replace(r'^http\://[a-zA-Z0-9\-\.]+\.[a-zA-Z]{2,3}(/\S*)?$',
                                  'webaddress')

# Replace money symbols with 'moneysymb' (£ can by typed with ALT key + 156)
df["reviews"] = df["reviews"].str.replace(r'£|\$', 'dollers')
    
# Replace 10 digit phone numbers (formats include paranthesis, spaces, no spaces, dashes) with 'phonenumber'
df["reviews"] = df["reviews"].str.replace(r'^\(?[\d]{3}\)?[\s-]?[\d]{3}[\s-]?[\d]{4}$',
                                  'phonenumber')

    
# Replace numbers with 'numbr'
df["reviews"] = df["reviews"].str.replace(r'\d+(\.\d+)?', 'numbr')
# Remove punctuation
df["reviews"] = df["reviews"].str.replace(r'[^\w\d\s]', ' ')

# Replace whitespace between terms with a single space
df["reviews"] = df["reviews"].str.replace(r'\s+', ' ')

# Remove leading and trailing whitespace
df["reviews"] = df["reviews"].str.replace(r'^\s+|\s+?$', '')

In [91]:
df.head(10)

ratings                                            reviews
0       5  Loved the product it s an A grade in every sph...
1       5  The delivery was super fast So writing about t...
2       5  Bass is excellent numbroutofnumbr Sound is nic...
3       5  I was finding best earbuds since many year I u...
4       5  quality is good for this price its just that i...
5       5  I gifted to my girlfriend and her review is as...
6       4  Nyc product value of the money bass is NYS and...
7       5  Excellent quality super b I never seen this ty...
8       4  Altogether the sound is good but the case is n...
9       5  Its best in all but having a size issue It is ...

In [103]:
df["ratings"].value_counts()

5    15592
4     3898
Name: ratings, dtype: int64

Importing the required Libraries

In [92]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report

Machine Learning

In [93]:
tfidf = TfidfVectorizer()

In [94]:
x = tfidf.fit_transform(df["reviews"])
y = df["ratings"]

In [95]:
x.shape

(19490, 222)

Splitting the data into x shape

In [96]:
y.shape

(19490,)

Splitting the data into y shape

In [97]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=.20,random_state=43)

Checking train split to the data set

In [98]:
x_train.shape

(15592, 222)

From the above case, we have 5592 Rows & 222 Columns

In [99]:
clf = LinearSVC()
clf.fit(x_train,y_train)

LinearSVC()

In [100]:
y_pred = clf.predict(x_test)

In [101]:
y_pred

array(['5', '5', '4', ..., '5', '5', '5'], dtype=object)

Prediction value of y is given in the above case

In [102]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           4       1.00      1.00      1.00       767
           5       1.00      1.00      1.00      3131

    accuracy                           1.00      3898
   macro avg       1.00      1.00      1.00      3898
weighted avg       1.00      1.00      1.00      3898



From the above case, the Accuracy score of Support Vector Machine (SVM) is 1.0